Documenting Custom Fields (#312)
SAP/cloud-sdk: docs/java/features/odata/type-safe-client-odata-v2.mdx

A detailed list of what is available in the SAP Cloud SDK can be obtained from [the Javadoc](https://help.sap.com/doc/b579bf8578954412aea2b458e8452201/1.0/en-US/com/sap/cloud/sdk/datamodel/odata/helper/package-summary.html).
The functionality can also be discovered through the fluent API.

### Accessing Extensible or Custom Fields
The concept of **In-app Extensibility** allows users to introduce additional fields to an existing SAP S/4HANA service.
To consume and work with such custom fields in your application code, SAP Cloud SDK introduces the concept of custom fields.

You can set custom fields on an entity through the `.setCustomField` method and read the value in the entity using `.getCustomField(fieldName)`

```java
businessPartner.setCustomField("remarks", "");

> businessPartner.setCustomField("shoeSize", 10);

Reviewer: Personally I would feel that one example is sufficient.

Author: Done.

final String remarks = businessPartner.getCustomField("remarks");
final int shoeSize = businessPartner.getCustomField("shoeSize");
```

> If you don't know the name of the custom fields added by the customer, you can still use `getCustomFieldNames()` to obtain the set of custom fields introduced and use this set further on to pull values from the entity.

Reviewer: Since the initial customer question was also asking about queries and multi tenancy consider adding information on that as well:
1. how to select custom fields?
2. Do custom fields work with multi tenancy? (they do)

Author: I have added a section to document usage of custom fields with system query options.

To consume and work with such custom fields in your application code, SAP Cloud SDK introduces the concept of custom fields.

You can set custom fields on an entity through the `.setCustomField` method and read the value in the entity using `.getCustomField(fieldName)`

```java
businessPartner.setCustomField("shoeSize", 10);

final int shoeSize = businessPartner.getCustomField("shoeSize");
```
If you don't know the name of the custom fields added by the customer, you can still use `getCustomFieldNames()` to obtain the set of custom fields introduced and use this set further on to pull values from the entity.
For e.g.

```java
final Set<String> customfieldNames = businessPartner.getCustomFieldNames();

for(String customFieldName: customfieldNames){
    final Object customField = businessPartner.getCustomField(customFieldName);
 }
```

The accessors for custom fields respect multitenancy.
`getCustomFieldNames()`, for example, fetches only custom field names available in the context of the current tenant.

### Using Custom Fields in system query options
You can also pass custom fields to select or filter system query options:

> <Tabs groupId="customField" defaultValue="v2" values={[ { label: 'OData V2', value: 'v2', }, { label: 'OData V2 (Deprecated)', value: 'v2-deprecated', } ]}>

Reviewer: ```suggestion
<Tabs groupId="odataProtocol" defaultValue="v2" values={[ { label: 'OData V2', value: 'v2', }, { label: 'OData V2 (Deprecated)', value: 'v2-deprecated', } ]}>
```

This will align the tab with all the other tabs: If you select the deprecated variant here, it will be selected everywhere.

I think it's not necessary to document the deprecated variant here, but it doesn't hurt either

 }
```

> The accessors for custom fields respect multitenancy, for e.g. `getCustomFieldNames()` would only fetch custom field names available in the context of the current tenant.

Reviewer 2: ```suggestion
The accessors for custom fields respect multitenancy.
`getCustomFieldNames()`, for example, fetches only custom field names available in the context of the current tenant.
```

### Using Custom Fields in system query options

> You can also pass custom fields to select or filter system query options.

Reviewer 2: ```suggestion
You can also pass custom fields to select or filter system query options:
```

<Tabs groupId="customField" defaultValue="v2" values={[ { label: 'OData V2', value: 'v2', }, { label: 'OData V2 (Deprecated)', value: 'v2-deprecated', } ]}>

> For e.g.

Reviewer 2: ```suggestion
```

<TabItem value="v2-deprecated">

```java
service().getAllBusinessPartner()
         .select(new BusinessPartnerField<>("ShoeSize"),()->"AnotherCustomField")
         .filter(new BusinessPartnerField<>("Shoesize").eq(42))
         .execute(destination)
```
</TabItem>

<TabItem value="v2">

```java
service().getAllBusinessPartner()
         .select(new BusinessPartnerField<>("ShoeSize"),()->"AnotherCustomField")
         .filter(new BusinessPartnerField<>("Shoesize").eq(42))
         .executeRequest(destination)
```
</TabItem>
</Tabs>

> Please be aware that such an example can work for all tenants only if `ShoeSize`, `AnotherCustomField` custom fields has been defined by all tenants.

Reviewer 2: ```suggestion
Please be aware that such an example can work for all tenants only if `ShoeSize` and `AnotherCustomField` are custom fields that have been defined by all tenants.
```
